### PR TITLE
EASY-2800: Improvements to EMD->DDM mapping

### PR DESCRIPTION
**Pull request by lindareijnhoudt**
Part of EASY-2800

#### When applied it will...
* EASY-2804: jaartal groter dan 19 miljoen
* EASY-2801: Only one ddm:created

### CODE REVIEW

File: src/test/scala/nl/knaw/dans/easy/fedoratobag/DdmSpec.scala [hunk 1]
```diff
@@ -1044,7 +1044,37 @@ class DdmSpec extends TestSupportFixture with EmdSupport with AudienceSupport wi
     }
   }
 
-  it should "render an invalid number of dates created" in {
+  it should "parse dates into reasonable dates" in {
+    val emd = parseEmdContent(Seq(
+      emdTitle, emdCreator, emdDescription,
+          <emd:date>
+            <eas:created eas:scheme="W3CDTF" eas:format="MILLISECOND">20180223-01-01T00:10:34.000+01:00</eas:created>
+            <eas:available eas:scheme="W3CDTF" eas:format="DAY">2017-09-30T17:47:36.978+02:00</eas:available>
```
Comment (jo-pol) [inline src/test/scala/nl/knaw/dans/easy/fedoratobag/DdmSpec.scala:1052]: What about
```suggestion
            <eas:available eas:scheme="W3CDTF" eas:format="DAY">20170930-01-01T17:47:36.978+02:00</eas:available>
```

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 2]
```diff
@@ -62,8 +64,8 @@ object DDM extends DebugEnhancedLogging {
        { /* instructions for reuse not specified as such in EMD */ }
        { emd.getEmdCreator.getDcCreator.asScala.map(bs => <dc:creator>{ bs.getValue.trim }</dc:creator>) }
        { emd.getEmdCreator.getEasCreator.asScala.map(author => <dcx-dai:creatorDetails>{ toXml(author)} </dcx-dai:creatorDetails>) }
-       { dateCreated.map(node =>  <ddm:created>{ node.text }</ddm:created>) }
-       { dateAvailable.map(node =>  <ddm:available>{ node.text }</ddm:available>) }
+       { if (dateCreated.nonEmpty) dateCreated.toSeq.head.map(node =>  <ddm:created>{ parseDate(node.text) }</ddm:created>) }
+       { dateAvailable.map(node =>  <ddm:available>{ parseDate(node.text) }</ddm:available>) }
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:68]: Rather build `parseDate` into `getDateMap` and make `dateMap` a Map[String, Iterable[String]]): https://github.com/DANS-KNAW/easy-fedora-to-bag/blob/6c717c885b3420a1320c85e628821b5766afcc1d/src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala#L40-L46

Comment (jo-pol) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:68]: Didn't we decide to apply these correction in https://github.com/DANS-KNAW/easy-convert-bag-to-deposit/
like ABR to ABR+?

Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:68]: i want the bags produced by fedora-to-bag to be valid. These are not valid dates (it was botched up by the webui). The bags in the vault are already valid, so these problems don't exist there

File: src/test/scala/nl/knaw/dans/easy/fedoratobag/DdmSpec.scala (hunk 1 above)
Comment (lindareijnhoudt) [reply src/test/scala/nl/knaw/dans/easy/fedoratobag/DdmSpec.scala:1052]: the thing is that only when eas:format=MILLISECOND do these strange dates appear. The MILLISECOND was the last-resort for the poor-mans-parser in the EMD library. So 
```
<eas:available eas:scheme="W3CDTF" eas:format="DAY">20170930T17:47:36.978+02:00</eas:available>
```
is not valid EMD, but with format=MILLISECOND it is ...

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 3]
```diff
@@ -38,9 +40,9 @@ object DDM extends DebugEnhancedLogging {
     //    println(new EmdMarshaller(emd).getXmlString)
 
     val dateMap: Map[String, Iterable[Elem]] = getDateMap(emd)
-    val dateCreated = dateMap("created")
+    val dateCreated = dateMap("created").map(d => parseDate(d.text))
     val dateAvailable = {
-      val elems = dateMap("available")
+      val elems = dateMap("available").map(d => parseDate(d.text))
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:45]: What about

    ...{ fixDate(value.getValue) }</label>
    ...{ fixDate(value.getDateTime) }</label>
at https://github.com/DANS-KNAW/easy-fedora-to-bag/blob/6c717c885b3420a1320c85e628821b5766afcc1d/src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala#L280-L282

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 4]
```diff
@@ -62,8 +64,8 @@ object DDM extends DebugEnhancedLogging {
        { /* instructions for reuse not specified as such in EMD */ }
        { emd.getEmdCreator.getDcCreator.asScala.map(bs => <dc:creator>{ bs.getValue.trim }</dc:creator>) }
        { emd.getEmdCreator.getEasCreator.asScala.map(author => <dcx-dai:creatorDetails>{ toXml(author)} </dcx-dai:creatorDetails>) }
-       { dateCreated.map(node =>  <ddm:created>{ node.text }</ddm:created>) }
-       { dateAvailable.map(node =>  <ddm:available>{ node.text }</ddm:available>) }
+       { if (dateCreated.nonEmpty)  <ddm:created>{ dateCreated.toSeq.head }</ddm:created> }
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:67]: `.head` is not safe and causes an un-descriptive exception when nothing is available.

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 5]
```diff
@@ -62,8 +64,8 @@ object DDM extends DebugEnhancedLogging {
...
+       { dateAvailable.map(date =>  <ddm:available>{ date }</ddm:available>) }
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:68]: Does this give the result you want? I'd expect it returns something with `Node` as prefix.

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 6]
```diff
@@ -91,10 +93,13 @@ object DDM extends DebugEnhancedLogging {
        { emd.getEmdCoverage.getEasSpatial.asScala.map(toXml) }
        <dct:license xsi:type="dct:URI">{ toLicenseUrl(emd.getEmdRights) }</dct:license>
        { emd.getEmdLanguage.getDcLanguage.asScala.map(bs => <dct:language xsi:type={langType(bs)}>{ langValue(bs) }</dct:language>) }
+       { if (dateCreated.size > 1)  dateCreated.toSeq.tail.map(date => <dct:created>{ date }</dct:created>) }
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:96]: I think the if-clause is redundant.

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 4 above)
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:67]: doesn't the `nonEmpty` take care of that?

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 5 above)
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:68]: the `dateCreated` and `dateAvailable` no longer contain Nodes, they have been parsed already. I hope the changed variable name would hint at that

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 6 above)
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:96]: you're right

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 4 above)
Comment (jo-pol) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:67]:
```suggestion
       { dateCreated.headOption.map(node =>  <ddm:created>{ node.text }</ddm:created>) }
```

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 5 above)
Comment (jo-pol) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:68]: Overlooked the change from node to date.

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/Spatial.scala [hunk 7]
```diff
@@ -117,11 +127,21 @@ case class SpatialBox(scheme: Option[String],
...
     </dcx-gml:spatial>
   )
+
+  private lazy val guessSrsName: String = srsName match {
+    case null => {
+      if(Seq(sNorth, sSouth).map(p => p.toDouble).max > 289000)
+        SpatialNames.RD_SRS_NAME
+      else
+        SpatialNames.DEGREES_SRS_NAME
+    }
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/easy/fedoratobag/Spatial.scala:144]: Extract val for north/south (cmd-alt-v), then extract method for the if-then-else (cmd-alt-m). The second action will be applied above too, if not I might be overlooking a difference. Then inline north/south again (cmd-alt-n).

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 8]
```diff
@@ -91,10 +93,13 @@ object DDM extends DebugEnhancedLogging {
        { emd.getEmdCoverage.getEasSpatial.asScala.map(toXml) }
        <dct:license xsi:type="dct:URI">{ toLicenseUrl(emd.getEmdRights) }</dct:license>
        { emd.getEmdLanguage.getDcLanguage.asScala.map(bs => <dct:language xsi:type={langType(bs)}>{ langValue(bs) }</dct:language>) }
+       { if (dateCreated.size > 1) dateCreated.toSeq.tail.map(date => <dct:created>{ date }</dct:created>) }
```
[the remark this answers is not in the document]
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:96]: no, i want everything but the first. The first goes into the profile, the rest goes here

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala [hunk 9]
```diff
@@ -301,12 +306,23 @@ object DDM extends DebugEnhancedLogging {
...
+      d
+  }
+
   private def toRelationXml(key: String, rel: Relation): Elem = Try {
     {
       <label scheme={ relationType(rel) }
              href={ Option(rel.getSubjectLink).map(toHref).orNull }
              xml:lang={ Option(rel.getSubjectTitle).map(_.getLanguage).orNull }
-      >{ Option(rel.getSubjectTitle).map(_.getValue.trim).getOrElse("") }</label>
+      >{ Option(rel.getSubjectTitle).getOrElse(new BasicString()).getValue.toOption.map(_.trim).getOrElse(rel.getSubjectLink) }</label>
```
[the remark this answers is not in the document]
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:325]: like so?
```suggestion
      >{
        val optionTitle = Option(rel.getSubjectTitle).getOrElse(new BasicString()).getValue.toOption
        optionTitle.map(_.trim)
          .getOrElse(rel.getSubjectLink) 
        }</label></label>
```

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala:67]: i'm giving up on this date. feel free to improve upon it later :)

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/Spatial.scala (hunk 7 above)
Comment (lindareijnhoudt) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/Spatial.scala]: done

File: src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (jo-pol) [reply src/main/scala/nl/knaw/dans/easy/fedoratobag/DDM.scala]: between line 204 and 305

**Pull request merged**